File: sdk/go/provider/tempmail_cn.go
package provider

import (
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"

	"github.com/gorilla/websocket"
)

const tempmailCNDefaultHost = "tempmail.cn"

var tempmailCNVersions = []int{4, 3}

var tempmailCNRegMu sync.Mutex
var tempmailCNBoxes = make(map[string]*tempmailCNBox)

type tempmailCNBox struct {
	mu      sync.Mutex
	emails  []NormEmail
	seenIDs map[string]struct{}
	started bool
}

func tempmailCNGetBox(email string) *tempmailCNBox {
	key := strings.ToLower(strings.TrimSpace(email))
	tempmailCNRegMu.Lock()
	defer tempmailCNRegMu.Unlock()
	b := tempmailCNBoxes[key]
	if b == nil {
		b = &tempmailCNBox{seenIDs: make(map[string]struct{})}
		tempmailCNBoxes[key] = b
	}
	return b
}

func tempmailCNNormalizeHost(domain *string) string {
	if domain == nil || strings.TrimSpace(*domain) == "" {
		return tempmailCNDefaultHost
	}
	raw := strings.TrimSpace(*domain)
	host := raw
	if strings.HasPrefix(strings.ToLower(host), "http://") || strings.HasPrefix(strings.ToLower(host), "https://") {
		if u, err := url.Parse(host); err == nil && u.Host != "" {
			host = u.Host
		}
	} else if strings.Contains(host, "/") {
		if u, err := url.Parse("https://" + host); err == nil && u.Host != "" {
			host = u.Host
		} else {
			host = strings.Split(host, "/")[0]
		}
	}
	host = strings.Trim(host, ".")
	if at := strings.LastIndex(host, "@"); at >= 0 {
		host = host[at+1:]
	}
	if host == "" {
		return tempmailCNDefaultHost
	}
	return host
}

func tempmailCNParseAddress(email string) (local string, host string, err error) {
	trimmed := strings.TrimSpace(email)
	at := strings.Index(trimmed, "@")
	if at <= 0 || at >= len(trimmed)-1 {
		return "", "", fmt.Errorf("tempmail-cn: invalid email address")
	}
	hostPart := trimmed[at+1:]
	return trimmed[:at], tempmailCNNormalizeHost(&hostPart), nil
}

func tempmailCNEventPacket(event string, payload interface{}) ([]byte, error) {
	b, err := json.Marshal([]interface{}{event, payload})
	if err != nil {
		return nil, err
	}
	return append([]byte("42"), b...), nil
}

func tempmailCNWriteEvent(conn *websocket.Conn, event string, payload interface{}) error {
	b, err := tempmailCNEventPacket(event, payload)
	if err != nil {
		return err
	}
	return conn.WriteMessage(websocket.TextMessage, b)
}

func tempmailCNParseEvent(packet string) (string, interface{}, bool) {
	if !strings.HasPrefix(packet, "42") {
		return "", nil, false
	}
	var arr []json.RawMessage
	if err := json.Unmarshal([]byte(packet[2:]), &arr); err != nil || len(arr) == 0 {
		return "", nil, false
	}
	var event string
	if err := json.Unmarshal(arr[0], &event); err != nil {
		return "", nil, false
	}
	if len(arr) < 2 {
		return event, nil, true
	}
	var payload interface{}
	if err := json.Unmarshal(arr[1], &payload); err != nil {
		return "", nil, false
	}
	return event, payload, true
}

func tempmailCNDial(host string) (*websocket.Conn, error) {
	var lastErr error
	for _, version := range tempmailCNVersions {
		u := url.URL{Scheme: "wss", Host: host, Path: "/socket.io/"}
		q := u.Query()
		q.Set("EIO", fmt.Sprintf("%d", version))
		q.Set("transport", "websocket")
		u.RawQuery = q.Encode()

		hdr := http.Header{}
		hdr.Set("Origin", "https://"+host)
		hdr.Set("Referer", "https://"+host+"/")
		hdr.Set("User-Agent", GetCurrentUA())
		hdr.Set("Accept-Language", "zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6")

		d := websocket.Dialer{HandshakeTimeout: 15 * time.Second}
		conn, _, err := d.Dial(u.String(), hdr)
		if err != nil {
			lastErr = err
			continue
		}
		if err := tempmailCNHandshake(conn); err != nil {
			conn.Close()
			lastErr = err
			continue
		}
		return conn, nil
	}
	if lastErr == nil {
		lastErr = fmt.Errorf("tempmail-cn: websocket handshake failed")
	}
	return nil, lastErr
}

func tempmailCNHandshake(conn *websocket.Conn) error {
	if err := conn.SetReadDeadline(time.Now().Add(15 * time.Second)); err != nil {
		return err
	}
	_, msg, err := conn.ReadMessage()
	if err != nil {
		return err
	}
	if len(msg) == 0 || msg[0] != '0' {
		return fmt.Errorf("tempmail-cn: unexpected open packet: %q", string(msg))
	}
	if err := conn.WriteMessage(websocket.TextMessage, []byte("40")); err != nil {
		return err
	}
	if err := conn.SetReadDeadline(time.Now().Add(1 * time.Second)); err != nil {
		return err
	}
	for {
		_, msg, err = conn.ReadMessage()
		if err != nil {
			if ne, ok := err.(interface{ Timeout() bool }); ok && ne.Timeout() {
				return conn.SetReadDeadline(time.Time{})
			}
			return err
		}
		packet := string(msg)
		if packet == "2" {
			if err := conn.WriteMessage(websocket.TextMessage, []byte("3")); err != nil {
				return err
			}
			continue
		}
		if strings.HasPrefix(packet, "40") {
			return conn.SetReadDeadline(time.Time{})
		}
	}
}

func tempmailCNString(v interface{}) string {
	if v == nil {
		return ""
	}
	s, ok := v.(string)
	if ok {
		return strings.TrimSpace(s)
	}
	return strings.TrimSpace(fmt.Sprintf("%v", v))
}

func tempmailCNStableID(raw map[string]interface{}, headers map[string]interface{}, recipient string) string {
	if id := tempmailCNString(raw["id"]); id != "" {
		return id
	}
	if id := tempmailCNString(raw["messageId"]); id != "" {
		return id
	}
	if id := tempmailCNString(headers["message-id"]); id != "" {
		return id
	}
	if id := tempmailCNString(headers["messageId"]); id != "" {
		return id
	}
	return fmt.Sprintf("%s\n%s\n%s\n%s",
		tempmailCNString(headers["from"]),
		tempmailCNString(headers["subject"]),
		tempmailCNString(headers["date"]),
		recipient,
	)
}

func tempmailCNFlatten(raw map[string]interface{}, recipient string) map[string]interface{} {
	headers, _ := raw["headers"].(map[string]interface{})
	if headers == nil {
		headers = map[string]interface{}{}
	}
	return map[string]interface{}{
		"id":          tempmailCNStableID(raw, headers, recipient),
		"from":        tempmailCNString(headers["from"]),
		"to":          recipient,
		"subject":     tempmailCNString(headers["subject"]),
		"text":        tempmailCNString(raw["text"]),
		"html":        tempmailCNString(raw["html"]),
		"date":        tempmailCNString(headers["date"]),
		"isRead":      false,
		"attachments": raw["attachments"],
	}
}

func tempmailCNRequestShortID(host string) (string, error) {
	conn, err := tempmailCNDial(host)
	if err != nil {
		return "", err
	}
	defer conn.Close()

	if err := tempmailCNWriteEvent(conn, "request shortid", true); err != nil {
		return "", err
	}
	if err := conn.SetReadDeadline(time.Now().Add(15 * time.Second)); err != nil {
		return "", err
	}
	defer conn.SetReadDeadline(time.Time{})

	for {
		_, msg, err := conn.ReadMessage()
		if err != nil {
			return "", err
		}
		packet := string(msg)
		if packet == "2" {
			if err := conn.WriteMessage(websocket.TextMessage, []byte("3")); err != nil {
				return "", err
			}
			continue
		}
		event, payload, ok := tempmailCNParseEvent(packet)
		if !ok || event != "shortid" {
			continue
		}
		id, ok := payload.(string)
		if ok && strings.TrimSpace(id) != "" {
			return id, nil
		}
	}
}

func TempmailCNGenerate(domain *string) (*CreatedMailbox, error) {
	host := tempmailCNNormalizeHost(domain)
	shortID, err := tempmailCNRequestShortID(host)
	if err != nil {
		return nil, err
	}
	return &CreatedMailbox{
		Channel: "tempmail-cn",
		Email:   shortID + "@" + host,
	}, nil
}

func tempmailCNReadLoop(email, local, host string, box *tempmailCNBox) {
	defer func() {
		box.mu.Lock()
		box.started = false
		box.mu.Unlock()
	}()

	conn, err := tempmailCNDial(host)
	if err != nil {
		return
	}
	defer conn.Close()

	if err := tempmailCNWriteEvent(conn, "set shortid", local); err != nil {
		return
	}

	for {
		_, msg, err := conn.ReadMessage()
		if err != nil {
			return
		}
		packet := string(msg)
		if packet == "2" {
			if err := conn.WriteMessage(websocket.TextMessage, []byte("3")); err != nil {
				return
			}
			continue
		}
		event, payload, ok := tempmailCNParseEvent(packet)
		if !ok || event != "mail" {
			continue
		}
		raw, ok := payload.(map[string]interface{})
		if !ok {
			continue
		}
		em := NormalizeMap(tempmailCNFlatten(raw, email), email)
		if strings.TrimSpace(em.ID) == "" {
			continue
		}
		box.mu.Lock()
		if _, dup := box.seenIDs[em.ID]; !dup {
			box.seenIDs[em.ID] = struct{}{}
			box.emails = append(box.emails, em)
		}
		box.mu.Unlock()
	}
}

func TempmailCNGetEmails(email string) ([]NormEmail, error) {
	local, host, err := tempmailCNParseAddress(email)
	if err != nil {
		return nil, err
	}
	box := tempmailCNGetBox(email)

	box.mu.Lock()
	needStart := !box.started
	if needStart {
		box.started = true
	}
	box.mu.Unlock()

	if needStart {
		go tempmailCNReadLoop(email, local, host, box)
		time.Sleep(80 * time.Millisecond)
	}

	box.mu.Lock()
	defer box.mu.Unlock()
	out := make([]NormEmail, len(box.emails))
	copy(out, box.emails)
	return out, nil
}

File: sdk/rust/src/providers/tempmail_cn.rs
/*!
 * tempmail.cn：按 public 前端的 Socket.IO 事件协议工作
 * - `request shortid` -> `shortid`
 * - `set shortid` 持续订阅 `mail`
 */

use std::collections::HashMap;
use std::sync::{Arc, Mutex, OnceLock};
use std::thread;
use std::time::Duration;

use serde_json::{json, Value};
use tungstenite::protocol::Message;

use crate::config::get_current_ua;
use crate::normalize::normalize_email;
use crate::types::{Channel, Email, EmailInfo};

const DEFAULT_HOST: &str = "tempmail.cn";
const SOCKET_IO_VERSIONS: &[u8] = &[4, 3];

type WsSocket = tungstenite::WebSocket<tungstenite::stream::MaybeTlsStream<std::net::TcpStream>>;

#[derive(Default)]
struct TempmailCnBox {
    emails: Vec<Email>,
    seen: HashMap<String, ()>,
    started: bool,
}

fn registry() -> &'static Mutex<HashMap<String, Arc<Mutex<TempmailCnBox>>>> {
    static REG: OnceLock<Mutex<HashMap<String, Arc<Mutex<TempmailCnBox>>>>> = OnceLock::new();
    REG.get_or_init(|| Mutex::new(HashMap::new()))
}

fn get_box(email: &str) -> Arc<Mutex<TempmailCnBox>> {
    let key = email.trim().to_lowercase();
    let mut g = registry().lock().unwrap();
    g.entry(key)
        .or_insert_with(|| Arc::new(Mutex::new(TempmailCnBox::default())))
        .clone()
}

fn normalize_host(domain: Option<&str>) -> String {
    let raw = domain.unwrap_or("").trim();
    if raw.is_empty() {
        return DEFAULT_HOST.to_string();
    }

    let mut host = raw.to_string();
    if host.starts_with("http://") || host.starts_with("https://") {
        if let Some(after) = host.split_once("://").map(|(_, rest)| rest) {
            host = after.to_string();
        }
    }
    if let Some((_, after_auth)) = host.rsplit_once('@') {
        host = after_auth.to_string();
    }
    if let Some((before_path, _)) = host.split_once('/') {
        host = before_path.to_string();
    }
    host = host.trim_matches('.').to_string();
    if host.is_empty() {
        DEFAULT_HOST.to_string()
    } else {
        host
    }
}

fn split_email(email: &str) -> Result<(String, String), String> {
    let trimmed = email.trim();
    let Some((local, host)) = trimmed.split_once('@') else {
        return Err("tempmail-cn: invalid email address".into());
    };
    if local.is_empty() || host.is_empty() {
        return Err("tempmail-cn: invalid email address".into());
    }
    Ok((local.to_string(), normalize_host(Some(host))))
}

fn socket_request(host: &str, version: u8) -> Result<tungstenite::http::Request<()>, String> {
    let ua = get_current_ua();
    let url = format!("wss://{host}/socket.io/?EIO={version}&transport=websocket");
    tungstenite::http::Request::builder()
        .uri(url)
        .header("Host", host)
        .header("Origin", format!("https://{host}"))
        .header("Referer", format!("https://{host}/"))
        .header("User-Agent", ua)
        .header("Accept-Language", "zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6")
        .body(())
        .map_err(|e| e.to_string())
}

fn socket_text(msg: Message) -> Option<String> {
    match msg {
        Message::Text(t) => Some(t.to_string()),
        Message::Binary(b) => String::from_utf8(b.to_vec()).ok(),
        Message::Close(_) => None,
        Message::Ping(_) | Message::Pong(_) | Message::Frame(_) => Some(String::new()),
    }
}

fn connect_socket(host: &str) -> Result<WsSocket, String> {
    let mut last_err = String::from("tempmail-cn: websocket handshake failed");
    for version in SOCKET_IO_VERSIONS {
        let request = match socket_request(host, *version) {
            Ok(r) => r,
            Err(e) => {
                last_err = e;
                continue;
            }
        };
        match tungstenite::connect(request) {
            Ok((mut socket, _)) => {
                let first = match socket.read() {
                    Ok(msg) => match socket_text(msg) {
                        Some(text) => text,
                        None => {
                            last_err = "tempmail-cn: websocket closed during handshake".to_string();
                            continue;
                        }
                    },
                    Err(e) => {
                        last_err = format!("tempmail-cn open packet: {e}");
                        continue;
                    }
                };
                if !first.starts_with('0') {
                    last_err = format!("tempmail-cn: unexpected open packet for EIO={version}: {first}");
                    continue;
                }
                if let Err(e) = socket.send(Message::Text("40".into())) {
                    last_err = format!("tempmail-cn connect namespace: {e}");
                    continue;
                }
                return Ok(socket);
            }
            Err(e) => {
                last_err = e.to_string();
            }
        }
    }
    Err(last_err)
}

fn send_event(socket: &mut WsSocket, event: &str, payload: Value) -> Result<(), String> {
    let packet = format!("42{}", json!([event, payload]));
    socket
        .send(Message::Text(packet.into()))
        .map_err(|e| e.to_string())
}

fn parse_event(packet: &str) -> Option<(String, Value)> {
    if !packet.starts_with("42") {
        return None;
    }
    let decoded: Vec<Value> = serde_json::from_str(&packet[2..]).ok()?;
    let event = decoded.first()?.as_str()?.to_string();
    let payload = decoded.get(1).cloned().unwrap_or(Value::Null);
    Some((event, payload))
}

fn value_string(v: Option<&Value>) -> String {
    let Some(v) = v else {
        return String::new();
    };
    match v {
        Value::String(s) => s.trim().to_string(),
        Value::Null => String::new(),
        _ => v.to_string().trim().to_string(),
    }
}

fn stable_message_id(raw: &Value, headers: &Value, recipient: &str) -> String {
    let candidates = [
        raw.get("id"),
        raw.get("messageId"),
        headers.get("message-id"),
        headers.get("messageId"),
    ];
    for candidate in candidates {
        let id = value_string(candidate);
        if !id.is_empty() {
            return id;
        }
    }
    format!(
        "{}\n{}\n{}\n{}",
        value_string(headers.get("from")),
        value_string(headers.get("subject")),
        value_string(headers.get("date")),
        recipient,
    )
}

fn flatten_mail(raw: &Value, recipient: &str) -> Value {
    let headers = raw.get("headers").unwrap_or(&Value::Null);
    json!({
        "id": stable_message_id(raw, headers, recipient),
        "from": value_string(headers.get("from")),
        "to": recipient,
        "subject": value_string(headers.get("subject")),
        "text": value_string(raw.get("text")),
        "html": value_string(raw.get("html")),
        "date": value_string(headers.get("date")),
        "isRead": false,
        "attachments": raw.get("attachments").cloned().unwrap_or(Value::Array(vec![])),
    })
}

fn request_shortid(host: &str) -> Result<String, String> {
    let mut socket = connect_socket(host)?;
    send_event(&mut socket, "request shortid", Value::Bool(true))?;

    loop {
        let packet = socket
            .read()
            .map_err(|e| e.to_string())
            .and_then(|msg| socket_text(msg).ok_or_else(|| "tempmail-cn: websocket closed before shortid".to_string()))?;
        if packet.is_empty() {
            continue;
        }
        if packet == "2" {
            socket
                .send(Message::Text("3".into()))
                .map_err(|e| e.to_string())?;
            continue;
        }
        let Some((event, payload)) = parse_event(&packet) else {
            continue;
        };
        if event == "shortid" {
            if let Some(id) = payload.as_str() {
                if !id.trim().is_empty() {
                    return Ok(id.trim().to_string());
                }
            }
        }
    }
}

fn ws_loop(email: String, local: String, host: String, arc: Arc<Mutex<TempmailCnBox>>) {
    let result = (|| -> Result<(), String> {
        let mut socket = connect_socket(&host)?;
        send_event(&mut socket, "set shortid", Value::String(local))?;

        loop {
            let packet = match socket.read() {
                Ok(msg) => match socket_text(msg) {
                    Some(t) => t,
                    None => break,
                },
                Err(e) => return Err(e.to_string()),
            };
            if packet.is_empty() {
                continue;
            }
            if packet == "2" {
                let _ = socket.send(Message::Text("3".into()));
                continue;
            }
            let Some((event, payload)) = parse_event(&packet) else {
                continue;
            };
            if event != "mail" {
                continue;
            }
            let em = normalize_email(&flatten_mail(&payload, &email), &email);
            if em.id.is_empty() {
                continue;
            }
            let mut inner = arc.lock().map_err(|e| e.to_string())?;
            if inner.seen.contains_key(&em.id) {
                continue;
            }
            inner.seen.insert(em.id.clone(), ());
            inner.emails.push(em);
        }

        Ok(())
    })();

    if let Err(e) = result {
        log::debug!("tempmail-cn ws loop end: {}", e);
    }
    if let Ok(mut inner) = arc.lock() {
        inner.started = false;
    }
}

fn ensure_ws(email: &str) -> Result<(), String> {
    let (local, host) = split_email(email)?;
    let arc = get_box(email);
    let need_start = {
        let mut inner = arc.lock().map_err(|e| e.to_string())?;
        if inner.started {
            false
        } else {
            inner.started = true;
            true
        }
    };

    if need_start {
        let email_owned = email.to_string();
        let arc_cloned = arc.clone();
        thread::spawn(move || ws_loop(email_owned, local, host, arc_cloned));
        thread::sleep(Duration::from_millis(80));
    }

    Ok(())
}

pub fn generate_email(domain: Option<&str>) -> Result<EmailInfo, String> {
    let host = normalize_host(domain);
    let shortid = request_shortid(&host)?;
    Ok(EmailInfo {
        channel: Channel::TempmailCn,
        email: format!("{}@{}", shortid, host),
        token: None,
        expires_at: None,
        created_at: None,
    })
}

pub fn get_emails(email: &str) -> Result<Vec<Email>, String> {
    ensure_ws(email)?;
    let arc = get_box(email);
    let inner = arc.lock().map_err(|e| e.to_string())?;
    Ok(inner.emails.clone())
}
